### ISSUE TITLE
npm install EPERM renaming package.json in cache

### ISSUE THREAD

**Issue by bryanmacfarlane** [npm/npm#9696]
installing ...
```
npm ERR! Windows_NT 6.3.9600
npm ERR! argv "C:\\Program Files\\nodejs\\\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install" "--loglevel" "error"
npm ERR! node v0.12.7
npm ERR! npm  v2.11.3
npm ERR! path C:\Users\vsoagentuser\AppData\Roaming\npm-cache\clone-stats\0.0.1\package\package.json
npm ERR! code EPERM
npm ERR! errno -4048

npm ERR! Error: EPERM, rename 'C:\Users\vsoagentuser\AppData\Roaming\npm-cache\clone-stats\0.0.1\package\package.json'
npm ERR!     at Error (native)
npm ERR!  { [Error: EPERM, rename 'C:\Users\vsoagentuser\AppData\Roaming\npm-cache\clone-stats\0.0.1\package\package.json']
npm ERR!   errno: -4048,
npm ERR!   code: 'EPERM',
npm ERR!   path: 'C:\\Users\\vsoagentuser\\AppData\\Roaming\\npm-cache\\clone-stats\\0.0.1\\package\\package.json',
npm ERR!   parent: 'vinyl' }
npm ERR!
npm ERR! Please try running this command again as root/Administrator.

npm ERR! Please include the following file with any support request:
npm ERR!     C:\test\npm-debug.log
```

npm-debug.log located here 
https://gist.github.com/bryanmacfarlane/cc7f49bfd0c651465121

How I Reproduced...  Our customers (visual studio online builds on azure VMs) also frequently reproduce

  1. Create Azure Windows 2012 Server VM (can't repro with fast local SSD dev box with Win10)
  2. Clear npm cache
  3. Use a package.json with a graph of many duplicated dependencies (bunch of gulp tasks is typical repro).  My package.json repro below

```json
{
  "name": "NpmRepro",
  "version": "1.0.0",
  "description": "NpmRepro",
  "main": "index.js",
  "author": "NpmReproer",
  "dependencies": {
    "del": "*",
    "gulp": "*",
    "gulp-angular-templatecache": "^1.7.0",
    "gulp-autoprefixer": "*",
    "gulp-concat": "*",
    "gulp-cli": "*",
    "gulp-if": "*",
    "gulp-less": "*",
    "gulp-minify-css": "*",
    "gulp-minify-html": "*",
    "gulp-ng-annotate": "*",
    "gulp-rev-easy": "*",
    "gulp-sourcemaps": "*",
    "gulp-typescript": "*",
    "gulp-uglify": "*",
    "gulp-useref": "*",
    "lazypipe": "*",
    "yargs": "*"
  }
}
```
With the repro steps I listed along with the package.json us and our customers can repro about 50% of the time.

Passing --force so it always gets from the server (and avoids the cache?) is the workaround.

**Comment by bryanmacfarlane** [comment]
What's the status on this.  Microsoft has plenty of customers using npm on azure VMs failing with this race condition on slower I/O machines.  We expect it to be slower but to succeed.  The repro above on slower I/O machines (even on other IAAS services or even local VMs) repros.

**Comment by joaomoreno** [comment]
We ([Microsoft/vscode](github.com/Microsoft/vscode)) are hitting this daily.

**Comment by AMoghrabi** [comment]
I am also having this issue on our team city build agents. Running a build has a 50% chance of failing with the mentioned error. I'm looking forward to a fix. Our agents are running Windows 7.

**Comment by LeviBotelho** [comment]
+1 Getting this error on an Azure VM that I'm using to build my Node project.

**Comment by joaocgreis** [comment]
This always reproduces in npm 2, using node 0.12.7, node 4.2.2 and even upgrading npm to 2.14.13. Does not seem to reproduce in node 5.1 (npm 3.3).

Possibly a duplicate of https://github.com/npm/npm/issues/7885 , but there npm is run multiple times simultaneously to install the same package. @artemsvirskyi thanks for https://github.com/artemsvirskyi/multiple-npm-install2 .

I can reproduce this on a local VM just by lowering CPU resources, either make it only one processor or using a CPU limit of 10%. Running as Administrator as suggested by npm does not help.

The error comes from the `write-file-atomic` package, when used here: https://github.com/npm/npm/blob/v2.14.13/lib/cache.js#L348 . Inspecting `write-file-atomic`, the EPERM error happens many more times, but is ignored in other uses of the package (like here: https://github.com/npm/npm/blob/v2.14.13/lib/cache/caching-client.js#L183 ).

**Comment by onezeronine** [comment]
+1 We are hitting this on our VSO hosted build. The commands ``npm set registry`` and ``npm cache clean`` does not work as some suggested.

**Comment by dmccaffery** [comment]
I am hitting this as well in AppVeyor.

Full build trace for funzies:
https://ci.appveyor.com/project/pulsebridge/condo/build/1.0.0.72

**Comment by Download** [comment]
Yes I got the same issue as well. It said on another issue that changing the cache folder might help, and it did, but the issue keeps coming back.

**Comment by bryanmacfarlane** [comment]
Workarounds that we've found:
  - use npm3
  - use --force flag (it bypasses the cache where the issue is)
  - use a VM with faster I/O (reduces probability of hitting it)

**Comment by Download** [comment]
```sh
c:\ws\react-mdl-naughty>npm install
npm WARN package.json react-mdl-naughty@0.1.0 No repository field.
npm WARN package.json react-mdl-naughty@0.1.0 No README data
npm WARN optional dep failed, continuing fsevents@1.0.6
npm ERR! Windows_NT 10.0.10586
npm ERR! argv "C:\\nodejs\\node.exe" "C:\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install"
npm ERR! node v4.2.3
npm ERR! npm  v2.14.7
npm ERR! path C:\npm\cache\js-tokens\1.0.2\package\package.json.04a80548edd35b2fb60416146f1bce47
npm ERR! code EPERM
npm ERR! errno -4048
npm ERR! syscall rename

npm ERR! Error: EPERM: operation not permitted, rename 'C:\npm\cache\js-tokens\1.0.2\package\package.json.04a80548edd35b2fb60416146f1bce47' -> 'C:\npm\cache\js-tokens\1.0.2\package\package.json'
npm ERR!     at Error (native)
npm ERR!  { [Error: EPERM: operation not permitted, rename 'C:\npm\cache\js-tokens\1.0.2\package\package.json.04a80548edd35b2fb60416146f1bce47' -> 'C:\npm\cache\js-tokens\1.0.2\package\package.json']
npm ERR!   errno: -4048,
npm ERR!   code: 'EPERM',
npm ERR!   syscall: 'rename',
npm ERR!   path: 'C:\\npm\\cache\\js-tokens\\1.0.2\\package\\package.json.04a80548edd35b2fb60416146f1bce47',
npm ERR!   dest: 'C:\\npm\\cache\\js-tokens\\1.0.2\\package\\package.json',
npm ERR!   parent: 'babel-code-frame' }
npm ERR!
npm ERR! Please try running this command again as root/Administrator.

npm ERR! Please include the following file with any support request:
npm ERR!     c:\ws\react-mdl-naughty\npm-debug.log

c:\ws\react-mdl-naughty>npm --version
2.14.7

c:\ws\react-mdl-naughty>node --version
v4.2.3
```

Sorry, I forgot to save the `npm-debug.log` file and on the next invocation to npm it got erased. I will attach it when I encounter this problem again.

**Comment by Download** [comment]
Oh, I found that deleting the folder it mentions helps. It succeeded on the next try.

**Comment by bryanmacfarlane** [comment]
@Download - have you tried the workarounds I mentioned above?

**Comment by Download** [comment]
The first and third workarounds are really too much of a burden. The `--force` I will try next time it occurs.

**Comment by conechan** [comment]
I met this issue. And I tried to use `--force`, but many other errors come out.

**Comment by bobbyg603** [comment]
+1, was not able to fix with --force, unfortunately upgrading the disk on our ESXi is not an option.

**Comment by devcube** [comment]
I also get this error on the VSO hosted build. Adding --force as argument seems to help.

**Comment by Download** [comment]
Force did not work for me. Fastest way for me is open the folder it complains about and delete it, then try again... Maybe NPM should try to implement the same procedure?

**Comment by joaocgreis** [comment]
This error happens because Windows can't replace a file atomically and npm assumes it can as unix does.

https://github.com/npm/npm/pull/10940 is waiting review in npm. Meanwhile, if someone can apply that manually to the installed npm (change `er` to `null`, thus ignoring the error, in line 350 of `lib/cache.js`), it would be useful to know if it works and everything installs as expected.

**Comment by astroanu** [comment]
there is one method you can do to get through this. I'm on a windows environment, but i use vagrant to simulate ubuntu development environment. running npm through vagrant solved this. it's a lot work, but works

**Comment by Download** [comment]
*knocks on wood*

I haven't had this happen anymore since I upgraded to Node 5.4.0 / NPM 3.3.12. If it does happen again I will report back here, but seeing how frequently it used to happen I'm starting to believe this is actually fixed in latest versions...

**Comment by bryanmacfarlane** [comment]
@download

Yes, from the beginning of this thread, we've identified the bug in npm 2.

Workarounds that we've found:

use npm3
use --force flag (it bypasses the cache where the issue is)
use a VM with faster I/O (reduces probability of hitting it)

**Comment by markelog** [comment]
> use npm3

We still see the issue in npm 3 -

> [npm cache clean] npm ERR! Windows_NT 6.1.7601
[18:43:34][npm cache clean] npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "--no-color" "cache" "clean"
[18:43:34][npm cache clean] npm ERR! node v5.3.0
[18:43:34][npm cache clean] npm ERR! npm  v3.3.12
[18:43:34][npm cache clean] npm ERR! path C:\Users\buildservice\AppData\Roaming\npm-cache
[18:43:34][npm cache clean] npm ERR! code EPERM
[18:43:34][npm cache clean] npm ERR! errno -4048
[18:43:34][npm cache clean] npm ERR! syscall rmdir
[18:43:34][npm cache clean] npm ERR! Error: EPERM: operation not permitted, rmdir 'C:\Users\buildservice\AppData\Roaming\npm-cache'

**Comment by markelog** [comment]
btw, `--force` also doesn't work, i see other users confirmed this as well - https://github.com/Medium/phantomjs/issues/19#issuecomment-170280904

**Comment by Download** [comment]
@markelog Actually that is a related, but subtly different error:

> EPERM: operation not permitted, rmdir 

versus

> EPERM: operation not permitted, rename

But it's both operation not permitted. The weird thing is, that AFAIK there aren't actually any permission problems. I'm running the NPM commands from a command prompt that I'm running 'As Administrator' and when I browse to the files myself, I can just delete it without issue. But Windows is known to have these quirks where it reports things as permission denied when they are actually due to locking etc.

Probably it can be solved by catching the error and retrying after a 100ms or something silly like that.

**Comment by dmccaffery** [comment]
@markelog nothing works reliably in my experience -- it just appears to work on occasion because the problem is so intermittent.

**Comment by bryanmacfarlane** [comment]
@markelog - you're hitting a different bug.  This bug is when they are trying to *rename* a file during cache writes and if you have a deep dependency tree with duplicates, one is trying to rename while another is renaming.  On windows they are using an api which is not atomic.  @joaocgreis root caused it above.

The signature of this bug is a rename of the package in the cache folder

npm ERR! Error: EPERM: operation not permitted, rename 'C:\npm\cache\js-tokens\1.0.2\package\package.json.04a80548edd35b2fb60416146f1bce47' -> 'C:\npm\cache\js-tokens\1.0.2\package\package.json'
npm ERR!     at Error (native)

**Comment by Download** [comment]
Hurray!! Great work guys!  :+1:

**Comment by markelog** [comment]
@iarna would you mind clarifying if your commit would also helps to our case - https://github.com/npm/npm/issues/9696#issuecomment-171343142

Or do we need to create different issue for that.

Also, thanks for responds everyone, they were very helpful

**Comment by joaocgreis** [comment]
@markelog your issue is not the same issue of this thread, so the fix will have no impact. From what you posted, it seems that npm cannot remove the directory `C:\Users\buildservice\AppData\Roaming\npm-cache`. Can you remove it manually? Does the user you run npm with have permissions to remove/create it? If this does not help you, perhaps the best is to open an issue for your problem.

**Comment by jakub-g** [comment]
For those of you wondering, the fix seems this was backported to 2.14.16

https://github.com/npm/npm/commit/5c0e4c45a29d774ab729e86044377d4e5e424252#diff-63fdc6a01d0ec0cbf9772723f2356e46

**Comment by Download** [comment]
I see the warning it now prints regularly; every time I see it I am glad you guys fixed this issue. Thank you!!

**Comment by howdoicomputer** [comment]
I'm getting a similar error. Related?

NPM: `3.8.0`
Windows: `10`
Node: `5.7.1`

```
npm ERR! Windows_NT 10.0.10586
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\drfra\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js" "install" "--force"
npm ERR! node v5.7.1
npm ERR! npm  v3.8.0
npm ERR! path C:\Users\drfra\OneDrive\Documents\GitHub\brigadehub\node_modules\ghost
npm ERR! code EPERM
npm ERR! errno -4048
npm ERR! syscall rename

npm ERR! Error: EPERM: operation not permitted, rename 'C:\Users\drfra\OneDrive\Documents\GitHub\brigadehub\node_modules\ghost' -> 'C:\Users\drfra\OneDrive\Documents\GitHub\brigadehub\node_modules\.ghost.DELETE'
npm ERR!     at moveAway (C:\Users\drfra\AppData\Roaming\npm\node_modules\npm\lib\install\action\finalize.js:38:5)
npm ERR!     at destStatted (C:\Users\drfra\AppData\Roaming\npm\node_modules\npm\lib\install\action\finalize.js:27:7)
npm ERR!     at FSReqWrap.oncomplete (fs.js:82:15)
npm ERR!
npm ERR! Error: EPERM: operation not permitted, rename 'C:\Users\drfra\OneDrive\Documents\GitHub\brigadehub\node_modules\ghost' -> 'C:\Users\drfra\OneDrive\Documents\GitHub\brigadehub\node_modules\.ghost.DELETE'
npm ERR!     at Error (native)
npm ERR!  { [Error: EPERM: operation not permitted, rename 'C:\Users\drfra\OneDrive\Documents\GitHub\brigadehub\node_modules\ghost' -> 'C:\Users\drfra\OneDrive\Documents\GitHub\brigadehub\node_modules\.ghost.DELETE'] parent: 'brigadehub' }
npm ERR!
npm ERR! Please try running this command again as root/Administrator.

npm ERR! Please include the following file with any support request:
npm ERR!     C:\Users\drfra\OneDrive\Documents\GitHub\brigadehub\npm-debug.log
```

**Comment by joaocgreis** [comment]
Node.js v4.4.0 was released with npm 2.14.20, including the fix for this issue! http://nodejs.org/dist/v4.4.0/

**Comment by adriaanthomas** [comment]
It looks like we're still hitting this on npm@2.14.20;
```
2016-03-25T00:12:37.6515576Z npm ERR! Windows_NT 6.3.9600
2016-03-25T00:12:37.6515576Z npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install"
2016-03-25T00:12:37.6515576Z npm ERR! node v4.4.0
2016-03-25T00:12:37.6515576Z npm ERR! npm  v2.14.20
2016-03-25T00:12:37.6515576Z npm ERR! path D:\npm-cache\mkdirp\0.5.1\package.tgz.1638249806
2016-03-25T00:12:37.6515576Z npm ERR! code EPERM
2016-03-25T00:12:37.6515576Z npm ERR! errno -4048
2016-03-25T00:12:37.6515576Z npm ERR! syscall rename
2016-03-25T00:12:37.6515576Z npm ERR! Error: EPERM: operation not permitted, rename 'D:\npm-cache\mkdirp\0.5.1\package.tgz.1638249806' -> 'D:\npm-cache\mkdirp\0.5.1\package.tgz'
2016-03-25T00:12:37.6515576Z npm ERR!     at Error (native)
2016-03-25T00:12:37.6515576Z npm ERR!  { [Error: EPERM: operation not permitted, rename 'D:\npm-cache\mkdirp\0.5.1\package.tgz.1638249806' -> 'D:\npm-cache\mkdirp\0.5.1\package.tgz']
2016-03-25T00:12:37.6515576Z npm ERR!   errno: -4048,
2016-03-25T00:12:37.6515576Z npm ERR!   code: 'EPERM',
2016-03-25T00:12:37.6515576Z npm ERR!   syscall: 'rename',
2016-03-25T00:12:37.6515576Z npm ERR!   path: 'D:\\npm-cache\\mkdirp\\0.5.1\\package.tgz.1638249806',
2016-03-25T00:12:37.6515576Z npm ERR!   dest: 'D:\\npm-cache\\mkdirp\\0.5.1\\package.tgz',
2016-03-25T00:12:37.6515576Z npm ERR!   parent: 'ssb2b' }
2016-03-25T00:12:37.6671853Z npm ERR! 
2016-03-25T00:12:37.6671853Z npm ERR! Please try running this command again as root/Administrator.
2016-03-25T00:12:37.6671853Z npm ERR! Please include the following file with any support request:
2016-03-25T00:12:37.6671853Z npm ERR!     D:\work\4\s\npm-debug.log
```

`npm-debug.log` is [here](https://gist.github.com/adriaanthomas/d22ccdcafab45739e337). This was an `npm install --force` with an empty cache directory (as the debug log shows), but clearly that did not help this time.

Clearing the cache sometimes helps, but just as often it will fail again on some package (not always the same one).

**Comment by adriaanthomas** [comment]
Oops, this is a different issue (not `package.json` related). I'll create a new issue.

**Comment by Daniel15** [comment]
I'm still seeing this on npm v2.15.0.

```
  npm ERR! argv "C:\\Program Files (x86)\\nodejs\\node.exe" "C:\\Program Files (x86)\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install"
  npm ERR! node v4.4.2
  npm ERR! npm  v2.15.0
  npm ERR! path C:\Users\appveyor\AppData\Roaming\npm-cache\core-js\1.2.6\package.tgz.3538883293
  npm ERR! code EPERM
  npm ERR! errno -4048
  npm ERR! syscall rename
  
EXEC : npm ERR! error : EPERM: operation not permitted, rename 'C:\Users\appveyor\AppData\Roaming\npm-cache\core-js\1.2.6\package.tgz.3538883293' -> 'C:\Users\appveyor\AppData\Roaming\npm-cache\core-js\1.2.6\package.tgz' [C:\projects\react-net\build.proj]
 
  npm ERR!     at Error (native)
  npm ERR!  { [Error: EPERM: operation not permitted, rename 'C:\Users\appveyor\AppData\Roaming\npm-cache\core-js\1.2.6\package.tgz.3538883293' -> 'C:\Users\appveyor\AppData\Roaming\npm-cache\core-js\1.2.6\package.tgz']
  npm ERR!   errno: -4048,
  npm ERR!   code: 'EPERM',
  npm ERR!   syscall: 'rename',
  npm ERR!   path: 'C:\\Users\\appveyor\\AppData\\Roaming\\npm-cache\\core-js\\1.2.6\\package.tgz.3538883293',
  npm ERR!   dest: 'C:\\Users\\appveyor\\AppData\\Roaming\\npm-cache\\core-js\\1.2.6\\package.tgz',
  npm ERR!   parent: 'babel-runtime' }
```

**Comment by tjwebb** [comment]
Also seeing this on npm 2.15.x.  `--force` does not fix the issue. Forcing npm3 is not a real "workaround" because users will be using whichever npm was bundled with node.

**Comment by benetis** [comment]
Been trying many different solutions with permissions, AV, versions - nothing worked until I figured I have scala server (JVM) running with "-Dsbt.jse.engineType=Node". Removing it solved the issue

**Comment by schleumer** [comment]
@benetis won the cookie. Play Framework was somehow reading my `package.json`.

**Comment by VishalGulati** [comment]
I am facing issues deploying angular2 app on azure. Could anyone please suggest what could be causing the error. I read some comments on running npm cache clean but dont know where to run it. Please refer my issue here. http://stackoverflow.com/questions/40241392/postinstall-command-in-package-json-fails-with-error-windows-nt-6-2-9200-on-azur

**Comment by h3rb** [comment]
This happens to me when using GIT BASH and ATOM editor.  If I have the project folder open in ATOM, npm generates npm WARN saveError EPERM: operation not permitted, rename

In the course of discovering this, I tried to take a lot of advice I found here: 
* turning off real-time protection in Windows Defender (no help)
* creating windows defender exceptions for the npm, development, project, nodejs and other related folder trees (python27) (no help)
* running gitbash escalated as administrator (no help)
* running npm cache clear or npm cache verify (no help)

It just came down to the fact that the file was locked by Atom Editor. (But it was not loaded in Atom, it was just in the project folder that was loaded.) This can happen when WinSCP is accessing the folder or the file, or other editors have it loaded (Notepad++ maybe?  Komodo?  not sure).  Visual Studio somehow avoids the locking issue and gets a signal when a file has changed.  I may submit an issue to Atom.io
